javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -56,8 +56,22 @@
     final byte leadingOffset, trailingLength;
 
     private Vector(T[] leading, byte leadingOffset, VectorTree<T> middle, T[] trailing, byte trailingLength) {
+        this.leading = leading;
+        this.middle = middle;
+        this.trailing = trailing;
+        this.leadingOffset = leadingOffset;
+        this.trailingLength = trailingLength;
+
+        assert (leading.length <= branchingFactor()) && (trailing.length <= branchingFactor());
+        assert (leadingLength() == 0) || ((leadingOffset < leading.length) && (length() > 0));
+        assert (trailingLength == 0) || (leading.length > this.leadingOffset);
+        assert (middle.length() == 0) || (trailingLength > 0);
+        assert List.range(0, middle.length()).forAll(i -> middle.getLeaf(i).length == branchingFactor());
+    }
+
+    private static <T> Vector<T> normalized(T[] leading, byte leadingOffset, VectorTree<T> middle, T[] trailing, byte trailingLength) {
```

Comment:
strangely the size of the constructor made the algorithms a lot slower, even if all the below `if` statements were false :|

Separated it to a simple constructor and normalization method, used mainly by `drop` and `take`.